ledger/participant-integration-api/src/main/scala/platform/store/interning/StringInterningCache.scala

```diff
@@ -0,0 +1,111 @@
+    * @return a completion Future: if the cache is behind it will load the missing entries from persistence, and update the cache state
+    */
+  def updateStringInterningCache(lastStringInterningId: Int)(implicit
+      loggingContext: LoggingContext
+  ): Future[Unit]
+}
+
+/** Encapsulate the dependency to load a range of string-interning-entries from persistence
+  */
+trait LoadStringInterningEntries {
+  def apply(
+      fromExclusive: Int,
+      toInclusive: Int,
+  ): LoggingContext => Future[Iterable[(Int, String)]]
+}
+
+/** This StringInterningCache representation uses the prefixed raw representation internally similar to the persistence layer.
+  * Concurrent cache usage is optimized for reading:
+  * - The single, volatile reference enables non-synchronized access from all threads, accessing persistent-immutable datastructure
+  * - On the writing side it synchronizes (this usage is anyway expected) and maintains the immutable internal datastructure
+  */
+class PrefixingStringInterningCache(loadPrefixedEntries: LoadStringInterningEntries)
+    extends StringInterningCache {
+  @volatile private var raw: RawStringInterning = RawStringInterning.from(Nil)
+
+  private def rawAccessor: StringInterningAccessor[String] = new StringInterningAccessor[String] {
+    override def internalize(t: String): Int = raw.map(t)
+    override def tryInternalize(t: String): Option[Int] = raw.map.get(t)
+    override def externalize(id: Int): String = raw.idMap(id)
+    override def tryExternalize(id: Int): Option[String] = raw.idMap.get(id)
+  }
+
+  private val TemplatePrefix = "t|"
+  private val PartyPrefix = "p|"
+
+  override val templateId: StringInterningDomain[Ref.Identifier] =
+    StringInterningDomain.prefixing(
+      prefix = TemplatePrefix,
+      prefixedAccessor = rawAccessor,
+      to = Ref.Identifier.assertFromString,
+      from = _.toString,
+    )
+
+  override val party: StringInterningDomain[Ref.Party] =
+    StringInterningDomain.prefixing(
+      prefix = PartyPrefix,
+      prefixedAccessor = rawAccessor,
+      to = Ref.Party.assertFromString,
+      from = _.toString,
+    )
+
+  override def add(domainStringIterators: DomainStringIterators): Iterable[(Int, String)] =
+    synchronized {
+      val allPrefixedStrings =
+        domainStringIterators.parties.map(PartyPrefix + _) ++
+          domainStringIterators.templateIds.map(TemplatePrefix + _)
+      val newEntries = RawStringInterning.newEntries(
+        strings = allPrefixedStrings,
+        rawStringInterning = raw,
+      )
+      updateCache(newEntries)
+      newEntries
+    }
+
+  override def updateStringInterningCache(
+      lastStringInterningId: Int
+  )(implicit loggingContext: LoggingContext): Future[Unit] =
+    if (lastStringInterningId <= raw.lastId) {
+      Future.unit
+    } else {
+      loadPrefixedEntries(raw.lastId, lastStringInterningId)(loggingContext)
+        .map(updateCache)(DirectExecutionContext)
```

Comment:
I would prefer having synchronous update method with already fetched data instead of passing a Future loading it from persistence.

Reply:
Seems like a sensible choice. I'm ok with injecting a Future 👍